Minor speedup in fmpz_pow_ui

albinahlback opened the pull request:
Setting both old version and new version in files separated from the profiling code, the results were:
```
5 bits:      min 1.03x,    max 1.02x
10 bits:      min 1.08x,    max 1.09x
15 bits:      min 1.10x,    max 1.10x
20 bits:      min 1.10x,    max 1.08x
25 bits:      min 1.12x,    max 1.10x
30 bits:      min 1.08x,    max 1.13x
35 bits:      min 1.08x,    max 1.11x
40 bits:      min 1.10x,    max 1.09x
45 bits:      min 1.07x,    max 1.07x
50 bits:      min 1.07x,    max 1.13x
55 bits:      min 1.07x,    max 1.08x
60 bits:      min 1.07x,    max 1.06x
65 bits:      min 1.08x,    max 1.05x
70 bits:      min 1.07x,    max 1.05x
75 bits:      min 0.96x,    max 1.04x
80 bits:      min 1.03x,    max 1.03x
85 bits:      min 1.01x,    max 1.04x
90 bits:      min 1.04x,    max 1.04x
95 bits:      min 1.02x,    max 1.02x
100 bits:      min 1.01x,    max 1.01x
105 bits:      min 0.99x,    max 1.03x
110 bits:      min 1.00x,    max 1.01x
115 bits:      min 1.03x,    max 1.03x
120 bits:      min 0.99x,    max 0.99x
125 bits:      min 1.02x,    max 1.02x
130 bits:      min 1.01x,    max 1.00x
135 bits:      min 1.02x,    max 1.01x
140 bits:      min 1.01x,    max 1.01x
145 bits:      min 1.01x,    max 0.99x
150 bits:      min 1.05x,    max 0.98x
```
When the old version was in the profiling file while the new version was in a different file, I got:
```
5 bits:      min 0.90x,    max 0.90x
10 bits:      min 0.97x,    max 0.97x
15 bits:      min 1.00x,    max 0.99x
20 bits:      min 0.99x,    max 0.99x
25 bits:      min 1.00x,    max 1.00x
30 bits:      min 1.00x,    max 0.99x
35 bits:      min 1.00x,    max 1.00x
40 bits:      min 1.00x,    max 1.01x
45 bits:      min 0.99x,    max 1.00x
50 bits:      min 1.00x,    max 1.00x
55 bits:      min 1.00x,    max 1.00x
60 bits:      min 0.99x,    max 0.99x
65 bits:      min 1.00x,    max 1.00x
70 bits:      min 0.98x,    max 0.99x
75 bits:      min 0.99x,    max 0.99x
80 bits:      min 0.98x,    max 0.98x
85 bits:      min 0.99x,    max 0.99x
90 bits:      min 0.97x,    max 0.99x
95 bits:      min 0.98x,    max 0.98x
100 bits:      min 0.97x,    max 0.97x
105 bits:      min 0.96x,    max 0.98x
110 bits:      min 0.97x,    max 0.97x
115 bits:      min 0.97x,    max 0.98x
120 bits:      min 0.99x,    max 0.98x
125 bits:      min 0.96x,    max 0.97x
130 bits:      min 0.98x,    max 0.97x
135 bits:      min 0.98x,    max 0.96x
140 bits:      min 0.98x,    max 0.97x
145 bits:      min 0.99x,    max 0.99x
150 bits:      min 0.97x,    max 0.97x
```

albinahlback commented:
I have absolutely no idea why the tests gets timed out.

albinahlback commented:
I think it should work now.

fredrik-johansson commented:
I think we should be more careful about checking for overflow (code like `exp * bits <= SMALL_FMPZ_BITCOUNT_MAX` looks unsafe) and guaranteeing that fmpz_pow_ui aborts safely instead of possibly silently returning nonsense or corrupting the internal state.

fmpz_pow_ui is one of those functions that is certainly seeing public use and where you can very easily pass wrong input by accident.

albinahlback commented:
> I think we should be more careful about checking for overflow (code like `exp * bits <= SMALL_FMPZ_BITCOUNT_MAX` looks unsafe) and guaranteeing that fmpz_pow_ui aborts safely instead of possibly silently returning nonsense or corrupting the internal state.
> 
> fmpz_pow_ui is one of those functions that is certainly seeing public use and where you can very easily pass wrong input by accident.

Ah, yes. I added abortion to the code and modified the profiling code to not favor new version (i.e. up the exponent so that first `if` statement with higher probability evaluates to false). I updated the results above.

wbhart commented:
I appreciate your effort! But we should think of this one as an experiment, which was very interesting to try, but ultimately proved unsuccessful.
